libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/DialogStateManager.cs

```diff
@@ -42,7 +50,23 @@ public DialogStateManager(DialogContext dc)
 
         public bool IsReadOnly => true;
 
-        public object this[string key] { get => GetValue<object>(key, () => null); set => SetValue(key, value); }
+        public object this[string key]
+        {
+            get => GetValue<object>(key, () => null);
+            set
+            {
+                if (key.IndexOfAny(Separators) == -1)
+                {
+                    // Root is handled by SetMemory rather than SetValue
+                    var scope = GetMemoryScope(key) ?? throw new ArgumentOutOfRangeException(GetBadScopeMessage(key));
+                    scope.SetMemory(this.dialogContext, value);
+                }
```

Comment:
why did you have to add this?

Reply:
Because otherwise you end with an infinite loop.  Object.SetPath identifies it as a dictionary does [key] which invokes SetValue which calls Object.SetPath and so on.  Your original code had this out as well.

---
In reply to: [360174705](https://github.com/microsoft/botbuilder-dotnet/pull/2936#discussion_r360174705) [](ancestors = 360174705)